src/snowflake/connector/cache.py

```diff
@@ -0,0 +1,264 @@
+        if is_expired(t):
+            self._expiration(k)
+            self.__delitem(k)
+            raise KeyError
+        if should_record_hits:
+            self._hit(k)
+        return v
+
+    def __setitem(
+        self,
+        k: K,
+        v: V,
+    ) -> None:
+        """Non-locking version of __setitem__.
+
+        This should only be used by internal functions when already
+        holding self._lock.
+        """
+        self._cache[k] = CacheEntry(
+            expiry=now() + self._entry_lifetime,
+            entry=v,
+        )
+        self.telemetry["size"] = len(self._cache)
+
+    def __getitem__(
+        self,
+        k: K,
+    ) -> V:
+        """Returns an element if it hasn't expired yet in a thread-safe way."""
+        with self._lock:
+            return self.__getitem(k, should_record_hits=True)
+
+    def __setitem__(
+        self,
+        k: K,
+        v: V,
+    ) -> None:
+        """Inserts an element in a thread-safe way."""
+        with self._lock:
+            self.__setitem(k, v)
+
+    def __iter__(self) -> Iterator[K]:
+        return iter(self.keys())
+
+    def keys(self) -> list[K]:
+        return [k for k, _ in self.items()]
+
+    def items(self) -> list[tuple[K, V]]:
+        with self._lock:
+            values: list[tuple[K, V]] = []
+            for k in list(self._cache.keys()):
+                try:
+                    # TODO: this function could be further optimized by removing
+                    #  the need here to call __getitem here and call se;f._cache.keys()
+                    v = self.__getitem(k, should_record_hits=False)
+                    values.append((k, v))
+                except KeyError:
+                    continue
+        return values
+
+    def values(self) -> list[V]:
+        return [v for _, v in self.items()]
+
+    def get(
+        self,
+        k: K,
+        default: V | None = None,
+    ) -> V | None:
+        try:
+            return self[k]
+        except KeyError:
+            return default
+
+    def clear(self) -> None:
+        with self._lock:
+            self._cache.clear()
+            self._reset_telemetry()
+
+    def __delitem(
+        self,
+        key: K,
+    ) -> None:
+        """Non-locking version of __delitem__.
+
+        This should only be used by internal functions when already
+        holding self._lock.
+        """
+        del self._cache[key]
+        self.telemetry["size"] = len(self._cache)
+
+    def __delitem__(
+        self,
+        key: K,
+    ) -> None:
+        with self._lock:
+            self.__delitem(key)
+
+    def __contains__(
+        self,
+        key: K,
+    ) -> bool:
+        with self._lock:
+            if key in self._cache.keys():
+                try:
+                    self.__getitem(key, should_record_hits=True)
+                    return True
+                except KeyError:
+                    # Fall through
+                    pass
+        return False
```

Comment:
I think this could be simplified as:
```suggestion
        with self._lock:
            try:
                self.__getitem(key, should_record_hits=True)
                return True
            except KeyError:
                # Fall through
                return False
```